File: leancomfy/haskell/Lattice/Utils/ColorUtils.hs
{-# LANGUAGE StrictData #-}

{-|
Module      : Lattice.Utils.ColorUtils
Description : Color conversion and manipulation
Copyright   : (c) Lattice, 2026
License     : MIT

HSV, RGB, HSL conversions and hex parsing.
All functions handle edge cases safely.

Source: leancomfy/lean/Lattice/Utils/ColorUtils.lean
-}

module Lattice.Utils.ColorUtils
  ( -- * Color Types
    HSV(..)
  , HSL(..)
  , HexParseResult(..)
    -- * Color Space Conversions
  , hsvToRgb
  , rgbToHsv
  , hslToRgb
  , rgbToHsl
    -- * Hex Conversion
  , hexToRgb
  , rgbToHex
  , rgbaToHex
  , hsvToHex
  , hexToHsv
    -- * Color Manipulation
  , lerpRgb
  , getContrastColor
    -- * Standard Swatches
  , standardSwatches
  ) where

import Data.Char (digitToInt, intToDigit, isHexDigit, toLower)
import Data.Text (Text)
import qualified Data.Text as T
import Lattice.Primitives

--------------------------------------------------------------------------------
-- Color Types
--------------------------------------------------------------------------------

-- | HSV color (Hue 0-360, Saturation 0-1, Value 0-1)
data HSV = HSV
  { hsvH :: !FiniteFloat  -- 0-360
  , hsvS :: !UnitFloat    -- 0-1
  , hsvV :: !UnitFloat    -- 0-1
  } deriving (Eq, Show)

-- | HSL color (Hue 0-360, Saturation 0-1, Lightness 0-1)
data HSL = HSL
  { hslH :: !FiniteFloat  -- 0-360
  , hslS :: !UnitFloat    -- 0-1
  , hslL :: !UnitFloat    -- 0-1
  } deriving (Eq, Show)

-- | Hex parse result
data HexParseResult
  = HexOk !RGB
  | HexOkWithAlpha !RGBA
  | HexInvalid !Text
  deriving (Eq, Show)

--------------------------------------------------------------------------------
-- Helper Functions
--------------------------------------------------------------------------------

-- | Check if Double is finite
isFiniteDouble :: Double -> Bool
isFiniteDouble x = not (isNaN x) && not (isInfinite x)

-- | Clamp value to range
clamp :: Double -> Double -> Double -> Double
clamp minVal maxVal x = max minVal (min maxVal x)

-- | Normalize hue to [0, 360)
normalizeHue :: Double -> FiniteFloat
normalizeHue h =
  let wrapped = ((h `mod'` 360) + 360) `mod'` 360
  in if isFiniteDouble wrapped
     then FiniteFloat wrapped
     else FiniteFloat 0
  where
    mod' a b = a - b * fromIntegral (floor (a / b) :: Int)

-- | Clamp to unit range
clampUnit :: Double -> UnitFloat
clampUnit = UnitFloat . clamp 0 1

-- | Make finite float
mkFinite :: Double -> FiniteFloat
mkFinite x = if isFiniteDouble x then FiniteFloat x else FiniteFloat 0

--------------------------------------------------------------------------------
-- Color Space Conversions
--------------------------------------------------------------------------------

-- | Convert HSV to RGB
hsvToRgb :: HSV -> RGB
hsvToRgb (HSV (FiniteFloat h) (UnitFloat s) (UnitFloat v)) =
  let hNorm = let wrapped = ((h `mod'` 360) + 360) `mod'` 360 in wrapped
      c = v * s
      x = c * (1 - abs (((hNorm / 60) `mod'` 2) - 1))
      m = v - c
      (r', g', b')
        | hNorm < 60 = (c, x, 0)
        | hNorm < 120 = (x, c, 0)
        | hNorm < 180 = (0, c, x)
        | hNorm < 240 = (0, x, c)
        | hNorm < 300 = (x, 0, c)
        | otherwise = (c, 0, x)
      toChannel n = mkFinite $ fromIntegral (round ((n + m) * 255) :: Int)
  in RGB (toChannel r') (toChannel g') (toChannel b')
  where
    mod' a b = a - b * fromIntegral (floor (a / b) :: Int)

-- | Convert RGB to HSV
rgbToHsv :: RGB -> HSV
rgbToHsv (RGB (FiniteFloat r) (FiniteFloat g) (FiniteFloat b)) =
  let r' = r / 255
      g' = g / 255
      b' = b / 255
      maxC = maximum [r', g', b']
      minC = minimum [r', g', b']
      d = maxC - minC
      s = if maxC == 0 then 0 else d / maxC
      v = maxC
      h | d == 0 = 0
        | maxC == r' = ((g' - b') / d + (if g' < b' then 6 else 0)) * 60
        | maxC == g' = ((b' - r') / d + 2) * 60
        | otherwise = ((r' - g') / d + 4) * 60
  in HSV (normalizeHue h) (clampUnit s) (clampUnit v)

-- | Convert HSL to RGB
hslToRgb :: HSL -> RGB
hslToRgb (HSL (FiniteFloat h) (UnitFloat s) (UnitFloat l)) =
  let hNorm = let wrapped = ((h `mod'` 360) + 360) `mod'` 360 in wrapped
      c = (1 - abs (2 * l - 1)) * s
      x = c * (1 - abs (((hNorm / 60) `mod'` 2) - 1))
      m = l - c / 2
      (r', g', b')
        | hNorm < 60 = (c, x, 0)
        | hNorm < 120 = (x, c, 0)
        | hNorm < 180 = (0, c, x)
        | hNorm < 240 = (0, x, c)
        | hNorm < 300 = (x, 0, c)
        | otherwise = (c, 0, x)
      toChannel n = mkFinite $ fromIntegral (round ((n + m) * 255) :: Int)
  in RGB (toChannel r') (toChannel g') (toChannel b')
  where
    mod' a b = a - b * fromIntegral (floor (a / b) :: Int)

-- | Convert RGB to HSL
rgbToHsl :: RGB -> HSL
rgbToHsl (RGB (FiniteFloat r) (FiniteFloat g) (FiniteFloat b)) =
  let r' = r / 255
      g' = g / 255
      b' = b / 255
      maxC = maximum [r', g', b']
      minC = minimum [r', g', b']
      l = (maxC + minC) / 2
      d = maxC - minC
      (h, s)
        | maxC == minC = (0, 0)
        | otherwise =
            let s' = if l > 0.5 then d / (2 - maxC - minC) else d / (maxC + minC)
                h' | maxC == r' = ((g' - b') / d + (if g' < b' then 6 else 0)) * 60
                   | maxC == g' = ((b' - r') / d + 2) * 60
                   | otherwise = ((r' - g') / d + 4) * 60
            in (h', s')
  in HSL (normalizeHue h) (clampUnit s) (clampUnit l)

--------------------------------------------------------------------------------
-- Hex Conversion
--------------------------------------------------------------------------------

-- | Parse hex color string to RGB
hexToRgb :: Text -> HexParseResult
hexToRgb hex =
  let hex' = T.unpack $ T.dropWhile (== '#') hex
  in case length hex' of
    -- #RGB format
    3 | all isHexDigit hex' ->
        let [r, g, b] = map digitToInt hex'
            r' = r * 16 + r
            g' = g * 16 + g
            b' = b * 16 + b
        in HexOk $ RGB (mkFinite $ fromIntegral r')
                       (mkFinite $ fromIntegral g')
                       (mkFinite $ fromIntegral b')

    -- #RRGGBB format
    6 | all isHexDigit hex' ->
        let pairs = [(hex' !! 0, hex' !! 1), (hex' !! 2, hex' !! 3), (hex' !! 4, hex' !! 5)]
            [r, g, b] = map hexPair pairs
        in HexOk $ RGB (mkFinite $ fromIntegral r)
                       (mkFinite $ fromIntegral g)
                       (mkFinite $ fromIntegral b)

    -- #RRGGBBAA format
    8 | all isHexDigit hex' ->
        let pairs = [(hex' !! 0, hex' !! 1), (hex' !! 2, hex' !! 3),
                     (hex' !! 4, hex' !! 5), (hex' !! 6, hex' !! 7)]
            [r, g, b, a] = map hexPair pairs
        in HexOkWithAlpha $ RGBA (mkFinite $ fromIntegral r)
                                 (mkFinite $ fromIntegral g)
                                 (mkFinite $ fromIntegral b)
                                 (clampUnit $ fromIntegral a / 255)

    _ -> HexInvalid $ T.pack $ "Invalid hex: " ++ hex'
  where
    hexPair (c1, c2) = digitToInt c1 * 16 + digitToInt c2

-- | Convert RGB to hex string
rgbToHex :: RGB -> Text
rgbToHex (RGB (FiniteFloat r) (FiniteFloat g) (FiniteFloat b)) =
  T.pack $ '#' : concatMap toHex [r, g, b]
  where
    toHex n = let x = max 0 (min 255 (round n :: Int))
              in [intToDigit (x `div` 16), intToDigit (x `mod` 16)]

-- | Convert RGBA to hex string with alpha
rgbaToHex :: RGBA -> Text
rgbaToHex (RGBA (FiniteFloat r) (FiniteFloat g) (FiniteFloat b) (UnitFloat a)) =
  T.pack $ '#' : concatMap toHex [r, g, b, a * 255]
  where
    toHex n = let x = max 0 (min 255 (round n :: Int))
              in [intToDigit (x `div` 16), intToDigit (x `mod` 16)]

-- | Convert HSV to hex string
hsvToHex :: HSV -> Text
hsvToHex = rgbToHex . hsvToRgb

-- | Convert hex to HSV
hexToHsv :: Text -> Maybe HSV
hexToHsv hex = case hexToRgb hex of
  HexOk rgb -> Just $ rgbToHsv rgb
  HexOkWithAlpha (RGBA r g b _) -> Just $ rgbToHsv (RGB r g b)
  HexInvalid _ -> Nothing

--------------------------------------------------------------------------------
-- Color Manipulation
--------------------------------------------------------------------------------

-- | Linear interpolation between two RGB colors
lerpRgb :: RGB -> RGB -> UnitFloat -> RGB
lerpRgb (RGB (FiniteFloat r1) (FiniteFloat g1) (FiniteFloat b1))
        (RGB (FiniteFloat r2) (FiniteFloat g2) (FiniteFloat b2))
        (UnitFloat t) =
  let lerp a b = mkFinite $ fromIntegral (round (a + (b - a) * t) :: Int)
  in RGB (lerp r1 r2) (lerp g1 g2) (lerp b1 b2)

-- | Get contrasting text color (black or white) for background
getContrastColor :: RGB -> RGB
getContrastColor (RGB (FiniteFloat r) (FiniteFloat g) (FiniteFloat b)) =
  let luminance = (0.299 * r + 0.587 * g + 0.114 * b) / 255
  in if luminance > 0.5
     then RGB (FiniteFloat 0) (FiniteFloat 0) (FiniteFloat 0)      -- black
     else RGB (FiniteFloat 255) (FiniteFloat 255) (FiniteFloat 255) -- white

--------------------------------------------------------------------------------
-- Standard Swatches
--------------------------------------------------------------------------------

-- | Standard color swatch hex values
standardSwatches :: [Text]
standardSwatches =
  [ "#ff0000", "#ff8000", "#ffff00", "#80ff00"
  , "#00ff00", "#00ff80", "#00ffff", "#0080ff"
  , "#0000ff", "#8000ff", "#ff00ff", "#ff0080"
  , "#ffffff", "#c0c0c0", "#808080", "#404040", "#000000"
  ]

File: ComfyUI-Lattice-Compositor/ui/src/utils/colorUtils.ts
/**
 * Color Utility Functions
 * HSV, RGB, HSL conversions and hex parsing
 */

import { clamp } from "./arrayUtils";

export type RGB = [number, number, number];
export type HSV = [number, number, number];
export type HSL = [number, number, number];
export type RGBA = [number, number, number, number];

/**
 * Convert HSV to RGB
 * @param h Hue (0-360)
 * @param s Saturation (0-1)
 * @param v Value (0-1)
 * @returns RGB tuple (0-255 each)
 */
export function hsvToRgb(h: number, s: number, v: number): RGB {
  h = ((h % 360) + 360) % 360;
  const c = v * s;
  const x = c * (1 - Math.abs(((h / 60) % 2) - 1));
  const m = v - c;

  let r = 0,
    g = 0,
    b = 0;
  if (h < 60) {
    r = c;
    g = x;
    b = 0;
  } else if (h < 120) {
    r = x;
    g = c;
    b = 0;
  } else if (h < 180) {
    r = 0;
    g = c;
    b = x;
  } else if (h < 240) {
    r = 0;
    g = x;
    b = c;
  } else if (h < 300) {
    r = x;
    g = 0;
    b = c;
  } else {
    r = c;
    g = 0;
    b = x;
  }

  return [
    Math.round((r + m) * 255),
    Math.round((g + m) * 255),
    Math.round((b + m) * 255),
  ];
}

/**
 * Convert RGB to HSV
 * @param r Red (0-255)
 * @param g Green (0-255)
 * @param b Blue (0-255)
 * @returns HSV tuple [hue (0-360), saturation (0-1), value (0-1)]
 */
export function rgbToHsv(r: number, g: number, b: number): HSV {
  r /= 255;
  g /= 255;
  b /= 255;

  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const d = max - min;

  let h = 0;
  const s = max === 0 ? 0 : d / max;
  const v = max;

  if (d !== 0) {
    switch (max) {
      case r:
        h = ((g - b) / d + (g < b ? 6 : 0)) * 60;
        break;
      case g:
        h = ((b - r) / d + 2) * 60;
        break;
      case b:
        h = ((r - g) / d + 4) * 60;
        break;
    }
  }

  return [h, s, v];
}

/**
 * Convert HSL to RGB
 * @param h Hue (0-360)
 * @param s Saturation (0-1)
 * @param l Lightness (0-1)
 * @returns RGB tuple (0-255 each)
 */
export function hslToRgb(h: number, s: number, l: number): RGB {
  h = ((h % 360) + 360) % 360;
  const c = (1 - Math.abs(2 * l - 1)) * s;
  const x = c * (1 - Math.abs(((h / 60) % 2) - 1));
  const m = l - c / 2;

  let r = 0,
    g = 0,
    b = 0;
  if (h < 60) {
    r = c;
    g = x;
    b = 0;
  } else if (h < 120) {
    r = x;
    g = c;
    b = 0;
  } else if (h < 180) {
    r = 0;
    g = c;
    b = x;
  } else if (h < 240) {
    r = 0;
    g = x;
    b = c;
  } else if (h < 300) {
    r = x;
    g = 0;
    b = c;
  } else {
    r = c;
    g = 0;
    b = x;
  }

  return [
    Math.round((r + m) * 255),
    Math.round((g + m) * 255),
    Math.round((b + m) * 255),
  ];
}

/**
 * Convert RGB to HSL
 * @param r Red (0-255)
 * @param g Green (0-255)
 * @param b Blue (0-255)
 * @returns HSL tuple [hue (0-360), saturation (0-1), lightness (0-1)]
 */
export function rgbToHsl(r: number, g: number, b: number): HSL {
  r /= 255;
  g /= 255;
  b /= 255;

  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);
  const l = (max + min) / 2;
  let h = 0;
  let s = 0;

  if (max !== min) {
    const d = max - min;
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min);

    switch (max) {
      case r:
        h = ((g - b) / d + (g < b ? 6 : 0)) * 60;
        break;
      case g:
        h = ((b - r) / d + 2) * 60;
        break;
      case b:
        h = ((r - g) / d + 4) * 60;
        break;
    }
  }

  return [h, s, l];
}

/**
 * Convert hex color to RGB
 * 
 * System F/Omega proof: Explicit validation of hex string format
 * Type proof: hex ∈ string → RGB (non-nullable)
 * Mathematical proof: Hex string must be valid 3, 6, or 8 character format to parse RGB values
 * 
 * @param hex Hex color string (#RGB, #RRGGBB, or #RRGGBBAA)
 * @returns RGB tuple (throws error if invalid - wrap in try/catch for expected "invalid hex" case)
 */
export function hexToRgb(hex: string): RGB {
  hex = hex.replace(/^#/, "");

  // Handle short format #RGB
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }

  // Handle #RRGGBB or #RRGGBBAA
  if (hex.length === 6 || hex.length === 8) {
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);

    if (!Number.isNaN(r) && !Number.isNaN(g) && !Number.isNaN(b)) {
      return [r, g, b];
    }
  }

  // System F/Omega proof: Explicit validation of hex string format
  // Type proof: hex ∈ string → [number, number, number] (non-nullable)
  // Mathematical proof: Hex string must be valid 3, 6, or 8 character format to parse RGB values
  throw new Error(
    `[colorUtils] Cannot convert hex to RGB: Invalid hex string format. ` +
    `Hex string: "${hex}", length: ${hex.length}, expected: 3, 6, or 8 characters (after # removal). ` +
    `Hex string must match format #RGB, #RRGGBB, or #RRGGBBAA. ` +
    `Wrap in try/catch if "invalid hex" is an expected state.`
  );
}

/**
 * Convert hex color to RGBA
 * 
 * System F/Omega proof: Explicit validation of hex string format
 * Type proof: hex ∈ string → RGBA (non-nullable)
 * Mathematical proof: Hex string must be valid 3, 6, or 8 character format to parse RGBA values
 * 
 * @param hex Hex color string (#RGB, #RRGGBB, or #RRGGBBAA)
 * @returns RGBA tuple (throws error if invalid - wrap in try/catch for expected "invalid hex" case)
 */
export function hexToRgba(hex: string): RGBA {
  hex = hex.replace(/^#/, "");

  // Handle short format #RGB
  if (hex.length === 3) {
    hex = `${hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2]}ff`;
  }

  // Handle #RRGGBB (add alpha)
  if (hex.length === 6) {
    hex = `${hex}ff`;
  }

  if (hex.length === 8) {
    const r = parseInt(hex.slice(0, 2), 16);
    const g = parseInt(hex.slice(2, 4), 16);
    const b = parseInt(hex.slice(4, 6), 16);
    const a = parseInt(hex.slice(6, 8), 16) / 255;

    if (
      !Number.isNaN(r) &&
      !Number.isNaN(g) &&
      !Number.isNaN(b) &&
      !Number.isNaN(a)
    ) {
      return [r, g, b, a];
    }
  }

  // System F/Omega proof: Explicit validation of hex string format
  // Type proof: hex ∈ string → [number, number, number, number] (non-nullable)
  // Mathematical proof: Hex string must be valid 8 character format to parse RGBA values
  throw new Error(
    `[colorUtils] Cannot convert hex to RGBA: Invalid hex string format. ` +
    `Hex string: "${hex}", length: ${hex.length}, expected: 8 characters (after # removal). ` +
    `Hex string must match format #RRGGBBAA. ` +
    `Wrap in try/catch if "invalid hex" is an expected state.`
  );
}

/**
 * Convert RGB to hex color string
 * @param r Red (0-255)
 * @param g Green (0-255)
 * @param b Blue (0-255)
 * @returns Hex color string (#RRGGBB)
 */
export function rgbToHex(r: number, g: number, b: number): string {
  const toHex = (n: number) =>
    Math.max(0, Math.min(255, Math.round(n)))
      .toString(16)
      .padStart(2, "0");
  return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
}

/**
 * Convert RGBA to hex color string with alpha
 * @param r Red (0-255)
 * @param g Green (0-255)
 * @param b Blue (0-255)
 * @param a Alpha (0-1)
 * @returns Hex color string (#RRGGBBAA)
 */
export function rgbaToHex(r: number, g: number, b: number, a: number): string {
  const toHex = (n: number) =>
    Math.max(0, Math.min(255, Math.round(n)))
      .toString(16)
      .padStart(2, "0");
  return `#${toHex(r)}${toHex(g)}${toHex(b)}${toHex(a * 255)}`;
}

/**
 * Convert HSV to hex color string
 */
export function hsvToHex(h: number, s: number, v: number): string {
  const [r, g, b] = hsvToRgb(h, s, v);
  return rgbToHex(r, g, b);
}

/**
 * Convert hex to HSV
 * 
 * System F/Omega proof: Explicit validation of hex string via hexToRgb dependency
 * Type proof: hex ∈ string → HSV (non-nullable)
 * Mathematical proof: Hex string must be valid to convert to RGB, then to HSV
 */
export function hexToHsv(hex: string): HSV {
  // System F/Omega pattern: hexToRgb throws error for invalid hex - propagate
  const rgb = hexToRgb(hex);
  return rgbToHsv(rgb[0], rgb[1], rgb[2]);
}

/**
 * Parse any color string to RGB
 * Supports: hex, rgb(), rgba(), hsl(), hsla()
 * 
 * System F/Omega proof: Explicit validation of color string format
 * Type proof: color ∈ string → RGB (non-nullable)
 * Mathematical proof: Color string must match hex, rgb/rgba, or hsl/hsla format
 */
export function parseColor(color: string): RGB {
  color = color.trim().toLowerCase();

  // Hex
  if (color.startsWith("#")) {
    return hexToRgb(color);
  }

  // rgb() or rgba()
  const rgbMatch = color.match(/rgba?\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)/);
  if (rgbMatch) {
    return [
      parseInt(rgbMatch[1], 10),
      parseInt(rgbMatch[2], 10),
      parseInt(rgbMatch[3], 10),
    ];
  }

  // hsl() or hsla()
  const hslMatch = color.match(
    /hsla?\s*\(\s*(\d+)\s*,\s*(\d+)%?\s*,\s*(\d+)%?/,
  );
  if (hslMatch) {
    return hslToRgb(
      parseInt(hslMatch[1], 10),
      parseInt(hslMatch[2], 10) / 100,
      parseInt(hslMatch[3], 10) / 100,
    );
  }

  // System F/Omega proof: Explicit validation of color string format
  // Type proof: color ∈ string → [number, number, number] (non-nullable)
  // Mathematical proof: Color string must match hex, rgb/rgba, or hsl/hsla format
  throw new Error(
    `[colorUtils] Cannot parse color string: Invalid format. ` +
    `Color string: "${color}", expected formats: #RGB, #RRGGBB, rgb(r,g,b), rgba(r,g,b,a), hsl(h,s%,l%), hsla(h,s%,l%,a). ` +
    `Color string must match one of the supported CSS color formats. ` +
    `Wrap in try/catch if "invalid color" is an expected state.`
  );
}

/**
 * Linear interpolation between two colors
 */
export function lerpColor(color1: string, color2: string, t: number): string {
  const rgb1 = hexToRgb(color1);
  const rgb2 = hexToRgb(color2);

  if (!rgb1 || !rgb2) return color1;

  t = clamp(t, 0, 1);

  return rgbToHex(
    Math.round(rgb1[0] + (rgb2[0] - rgb1[0]) * t),
    Math.round(rgb1[1] + (rgb2[1] - rgb1[1]) * t),
    Math.round(rgb1[2] + (rgb2[2] - rgb1[2]) * t),
  );
}

/**
 * Get contrasting text color (black or white) for a background
 */
export function getContrastColor(bgColor: string): string {
  const rgb = hexToRgb(bgColor);
  if (!rgb) return "#ffffff";

  // Calculate luminance
  const luminance = (0.299 * rgb[0] + 0.587 * rgb[1] + 0.114 * rgb[2]) / 255;
  return luminance > 0.5 ? "#000000" : "#ffffff";
}

/**
 * Standard color swatches
 */
export const STANDARD_SWATCHES = [
  "#ff0000",
  "#ff8000",
  "#ffff00",
  "#80ff00",
  "#00ff00",
  "#00ff80",
  "#00ffff",
  "#0080ff",
  "#0000ff",
  "#8000ff",
  "#ff00ff",
  "#ff0080",
  "#ffffff",
  "#c0c0c0",
  "#808080",
  "#404040",
  "#000000",
];
